src/theory/uf/eq_proof.h

```diff
@@ -0,0 +1,338 @@
+   * @param p a pointer to a CDProof to store the conversion of this EqProof
+   * @param assumptions the assumptions (and variants) of the original
+   * EqProof. These are necessary to avoid cyclic proofs, which could be
+   * generated by creating transitivity steps for assumptions (which depend on
+   * themselves).
+   */
+  bool expandTransitivityForDisequalities(
+      Node conclusion,
+      std::vector<Node>& premises,
+      CDProof* p,
+      std::unordered_set<Node, NodeHashFunction>& assumptions) const;
+
+  /** Builds a transitivity chain from equalities to derive a conclusion
+   *
+   * Given an equality (= t1 tn), and a list of equalities premises, attempts to
+   * build a chain (= t1 t2) ... (= tn-1 tn) from premises. This is done in a
+   * greedy manner by finding one "link" in the chain at a time, updating the
+   * conclusion to be the next link and the premises by removing the used
+   * premise, and searching recursively.
+   *
+   * Consider for example
+   * - conclusion: (= t1 t4)
+   * - premises:   (= t4 t2), (= t2 t3), (= t2 t1), (= t3 t4)
+   *
+   * It proceeds by searching for t1 in an equality in the premises, in order,
+   * which is found in the equality (= t2 t1). Since t2 != t4, it attempts to
+   * build a chain with
+   * - conclusion (= t2 t4)
+   * - premises (= t4 t2), (= t2 t3), (= t3 t4)
+   *
+   * In the first equality it finds t2 and also t4, which closes the chain, so
+   * that premises is updated to (= t2 t4) and the method returns true. Since
+   * the recursive call was successful, the original conclusion (= t1 t4) is
+   * justified with (= t1 t2) plus the chain built in the recursive call,
+   * i.e. (= t1 t2), (= t2 t4).
+   *
+   * Note that not all the premises were necessary to build a successful
+   * chain. Moreover, note that building a successful chain can depend on the
+   * order in which the equalities are chosen. When a recursive call fails to
+   * close a chain, the chosen equality is dismissed and another is searched for
+   * among the remaining ones.
+   *
+   * This method assumes that premises does not contain reflexivity premises.
+   * This is without loss of generality since such premises are spurious for a
+   * transitivity step.
+   *
+   * @param conclusion the conclusion of the transitivity chain of equalities
+   * @param premises the list of equalities to build a chain with
+   * @return whether premises successfully build a transitivity chain for the
+   * conclusion
+   */
+  bool buildTransitivityChain(Node conclusion,
+                              std::vector<Node>& premises) const;
+
+  /** Reduce the a congruence EqProof into a transitivity matrix
+   *
+   * Given a congruence EqProof of (= (f a0 ... an-1) (f b0 ... bn-1)), reduce
+   * its justification into a matrix
+   *
+   *   [0]   -> p_{0,0} ... p_{m_0,0}
+   *   ...
+   *   [n-1] -> p_{0,n} ... p_{m_n-1,n-1}
+   *
+   * where f has arity n and each p_{0,i} ... p_{m_i, i} contains a transitivity
```

Comment:
Not "contains", I think this should say "for each i, .... is a transitivity chain".  I was confused why there were seemingly multiple proofs of the same thing for each range.

Reply:
I was trying to say that the chain could be minimized, i.e. not all equalities may be required for justifying the premise. I'll change to what you suggested to avoid confusion.